test-js/unit/pentaho/CustomContextVars.Spec.js

```
@@ -0,0 +1,113 @@
+
+  // Use alternate, promise-aware version of `it`.
+  var it = testUtils.itAsync;
+
+  describe("pentaho.CustomContextVars -", function() {
+
+    /**
+     * Mocks AMD with dummy default context variables.
+     */
+    function configDefaultContext(localRequire) {
+
+      localRequire.define("pentaho/contextVars", {
+        application: "APP",
+        user:        "USER",
+        theme:       "THEME",
+        locale:      "LOCALE"
+      });
+    }
+
+    it("is a function", function() {
+
+      return require.using(["pentaho/CustomContextVars"], configDefaultContext, function(CustomContextVars) {
+        expect(typeof CustomContextVars).toBe("function");
+      });
+    });
+
+    describe("new CustomContextVars(spec) -", function() {
+
+      it("should return a CustomContextVars instance", function() {
+
+        return require.using(["pentaho/CustomContextVars"], configDefaultContext, function(CustomContextVars) {
+          var contextVars = new CustomContextVars();
+          expect(contextVars instanceof CustomContextVars).toBe(true);
+        });
+      });
+
+      /**
+       * Tests that a context variable is properly implemented by pentaho.CustomContextVars.
+       *
+       * @param {string} varName - The name of the context variable's.
+       */
+      function testVariable(varName) {
+        describe(varName, function() {
+
+          it("should default to the default variable's value, when spec." + varName + " is not specified", function() {
+
+            return require.using([
+              "pentaho/contextVars",
+              "pentaho/CustomContextVars"
+            ], configDefaultContext, function(defaultContextVars, CustomContextVars) {
+
+              var contextVars = new CustomContextVars();
+              expect(contextVars[varName]).toBe(defaultContextVars[varName]);
+            });
+          });
+
+          it("should default to the default variable's value, when spec." + varName + " is empty", function() {
```

For instance... suppose I want to perform something impersonating a logged out user, creating a context with `{user: null}` value would default to the current user, not to "no user".
That probably isn't the current semantics of the field, or not even a real scenario, but generically speaking seems to be a valid use case for allowing `null` values.

Omitting fields in the spec (or explicitly setting them to `undefined`, but there's no need for that) is the way I see making sense to fallback to the default values. Falling back on invalid (wrong type) or empty values seem strange to me.